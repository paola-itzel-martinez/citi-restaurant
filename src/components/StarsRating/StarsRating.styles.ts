import styled from 'styled-components';
import { palette } from '../../styled';

export const Wrapper = styled.div`
  display: flex;
  gap: 2px;
`;

export const StarIcon = styled.i`
  display: inline-block;
  color: ${palette.YELLOW};
  overflow: hidden;
  width: 16px;
  text-overflow: clip;
`;
