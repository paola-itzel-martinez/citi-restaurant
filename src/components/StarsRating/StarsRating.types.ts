export interface StarProps {
  rating: number;
}
