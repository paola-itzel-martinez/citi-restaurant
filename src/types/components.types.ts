export enum ButtonVariant {
  CARD = 'CARD',
  MENU = 'MENU',
}

export enum DividerVariant {
  HORIZONTAL = 'HORIZONTAL',
  VERTICAL = 'VERTICAL',
}

export enum Portals {
  MODAL = 'MODAL',
}

export enum TypographyVariant {
  BUTTON = 'BUTTON',
  BUTTON_MENU = 'BUTTON_MENU',
  NORMAL = 'NORMAL',
  SMALL = 'SMALL',
  TITLE = 'TITLE',
}

export enum TypographyFamily {
  GENERAL = "'Roboto',sans-serif",
}

export type CSSRule = number | string;

export type CSSMarginRules = {
  m?: CSSRule;
  mx?: CSSRule;
  my?: CSSRule;
  mt?: CSSRule;
  mr?: CSSRule;
  mb?: CSSRule;
  ml?: CSSRule;
};

export type CSSPaddingRules = {
  p?: CSSRule;
  px?: CSSRule;
  py?: CSSRule;
  pt?: CSSRule;
  pr?: CSSRule;
  pb?: CSSRule;
  pl?: CSSRule;
};

export type CSSSpacingRules = {
  width?: CSSRule;
  minWidth?: CSSRule;
  maxWidth?: CSSRule;
  height?: CSSRule;
  minHeight?: CSSRule;
  maxHeight?: CSSRule;
};
