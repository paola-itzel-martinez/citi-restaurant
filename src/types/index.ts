export * from './components.types';
export * from './mediaQueries.types';
export * from './yelpAPI.types';
