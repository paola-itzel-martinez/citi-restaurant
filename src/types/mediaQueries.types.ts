export enum SCREEN_SIZE {
  DESKTOP = 'DESKTOP',
  TABLET = 'TABLET',
  MOBILE = 'MOBILE',
}
