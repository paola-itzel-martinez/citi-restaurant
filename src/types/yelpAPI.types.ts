export type Category = {
  title: string;
  alias: string;
};

export type Business = {
  id: string;
  categories: Category[];
  image_url: string;
  name: string;
  rating: number;
  url: string;
  price?: string;
};
