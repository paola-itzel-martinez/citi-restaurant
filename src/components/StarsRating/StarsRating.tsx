import React from "react";
import { StarIcon, Wrapper } from "./StarsRating.styles";
import { StarProps } from "./StarsRating.types";

const TOTAL_STARS = 5;

const StarsRating = ({ rating }: StarProps) => {
  const filledStars = [...Array(Math.floor(rating))];
  const emptyStars = [...Array(Math.floor(TOTAL_STARS - Math.ceil(rating)))];
  const hasHalfStar = rating % 1 !== 0;

  return (
    <Wrapper>
      {filledStars?.map((_, index) => (
        <StarIcon
          key={`filled-rating-${index}`}
          data-testid="filled-star"
          className="fas fa-star"
        />
      ))}

      {hasHalfStar && (
        <StarIcon
          key="half-star"
          data-testid="half-star"
          className="fas fa-star-half-alt"
        />
      )}

      {emptyStars.map((index) => (
        <StarIcon
          key={`empty-star-${index}`}
          data-testid="empty-star"
          className="far fa-star"
        />
      ))}
    </Wrapper>
  );
};

export default React.memo(StarsRating);
