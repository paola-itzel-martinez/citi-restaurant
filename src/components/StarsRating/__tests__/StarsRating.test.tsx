import React from "react";
import { render, screen } from "../../../../tests";
import { StarProps } from "../StarsRating.types";
import { StarsRating } from "..";

const GeneralComponent = (props: StarProps) => <StarsRating {...props} />;

describe("<StarsRating />", () => {
  it("Should have a specific css rules if has all stars filled", () => {
    const { container } = render(<GeneralComponent rating={5} />);

    expect(screen.getAllByTestId("filled-star")).toHaveLength(5);
    expect(container).toMatchSnapshot();
  });

  it("Should have a specific css rules if has a half star", () => {
    const { container } = render(<GeneralComponent rating={3.5} />);

    expect(screen.getAllByTestId("filled-star")).toHaveLength(3);
    expect(screen.getAllByTestId("half-star")).toHaveLength(1);
    expect(screen.getAllByTestId("empty-star")).toHaveLength(1);
    expect(container).toMatchSnapshot();
  });

  it("Should have a specific css rules if has all stars empty", () => {
    const { container } = render(<GeneralComponent rating={0} />);

    expect(screen.getAllByTestId("empty-star")).toHaveLength(5);
    expect(container).toMatchSnapshot();
  });
});
