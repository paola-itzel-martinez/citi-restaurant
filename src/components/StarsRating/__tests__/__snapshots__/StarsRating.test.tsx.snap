// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarsRating /> Should have a specific css rules if has a half star 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
}

.c1 {
  display: inline-block;
  color: #ffc104;
  overflow: hidden;
  width: 16px;
  text-overflow: clip;
}

<div>
  <div
    class="c0"
  >
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star-half-alt"
      data-testid="half-star"
    />
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
  </div>
</div>
`;

exports[`<StarsRating /> Should have a specific css rules if has all stars empty 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
}

.c1 {
  display: inline-block;
  color: #ffc104;
  overflow: hidden;
  width: 16px;
  text-overflow: clip;
}

<div>
  <div
    class="c0"
  >
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
    <i
      class="c1 far fa-star"
      data-testid="empty-star"
    />
  </div>
</div>
`;

exports[`<StarsRating /> Should have a specific css rules if has all stars filled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
}

.c1 {
  display: inline-block;
  color: #ffc104;
  overflow: hidden;
  width: 16px;
  text-overflow: clip;
}

<div>
  <div
    class="c0"
  >
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
    <i
      class="c1 fas fa-star"
      data-testid="filled-star"
    />
  </div>
</div>
`;
